### PR TITLE
Temporal mtls using azure keyvault certificates

### PR DESCRIPTION
**Pull request by Aymalla**
This PR is an initial implementation for the temporal MTLS cluster, worker, and helloworld workflows
based on temporalio samples
- [Temporal SSL Java Sample](https://github.com/temporalio/samples-java/blob/main/src/main/java/io/temporal/samples/ssl/SslEnabledWorker.java)
- [Temporal Server tls-sample](https://github.com/temporalio/samples-server/tree/main/tls/tls-simple)

Items to be reviewed
- Generatings the certificates
- Starting of Temporal cluster with mTLS enabled
- Starting a Temporal Worker
- Trigger Helloworld workflow
- The documents are clear

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,58 @@
...
+
+`**Note:** In the case of using the dev-container and Vscode, all dependencies are already installed.`
+
+## Get Started
+
+- Clone repository `git clone https://github.com/Aymalla/temporalio-mtls.git`.
+- An Azure Key Vault need to be created to store certificates.
+The caller also needs `Key Vault Certificates Officer` and `Key Vault Secrets User` RBAC roles
+to create, import certificates and download private keys.
+- Login to your subscription `az login`
```
Comment (phongcao) [inline README.md]: nit: missing `.` at the end.

File: src/main/java/com/temporal/samples/helloworld/Application.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+@SpringBootApplication
+@EnableAsync
+public class Application {
+	private Client temporalClient;
+
+	public Application(Client temporalClient) {
+		this.temporalClient = temporalClient;
+	}
+	public static void main(String[] args) {
```
Comment (phongcao) [inline src/main/java/com/temporal/samples/helloworld/Application.java:21]: nit: missing empty line above

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+temporal.server.certAuthorityName = tls-sample
+
+temporal.workflow.taskqueue = MAIN_TASK_QUEUE
+temporal.workflow.id = SamplesWorkflow
+
+# mTLS certificate paths
+temporal.tls.ca.certPath = deployment/certs/certs/ca.cert
+temporal.tls.client.certPath = deployment/certs/certs/client.pem
+temporal.tls.client.keyPath = deployment/certs/certs/client.key
+
```
Comment (phongcao) [inline src/main/resources/application.properties]: nit: delete redundant empty lines

File: src/test/java/com/temporal/samples/approval/HelloworldApplicationTests.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.temporal.workflows.registercredit;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class ApprovalApplicationTests {
```
Comment (phongcao) [inline src/test/java/com/temporal/samples/approval/HelloworldApplicationTests.java]: Change class name to Helloworld tests

File: temporal-mtls.md [hunk 5]
```diff
@@ -0,0 +1,341 @@
+# Setup and configure a temporal cluster with mTLS certificates using Azure Key-vault
```
Comment (phongcao) [inline temporal-mtls.md:1]: Please update this readme basing on my comments on your Medium blog post.

File: deployment/certs/generate-test-certs-keyvault.sh [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    # generate the CA private key
+    openssl genrsa -out $CERTS_DIR/$certName.key 4096
+    
+    # generate the CA certificate
+    openssl req -new -x509 -key $CERTS_DIR/$certName.key -sha256 -subj "/OU=Test CA Corporation/O=Test CA Corporation/L=London/S=Greater London/C=UK" -days 365 -out $CERTS_DIR/$certName.cert
+    
+    # generate the CA certificate .pfx file
+    openssl pkcs12 -export -out "$CERTS_DIR/$certName.pfx" -inkey "$CERTS_DIR/$certName.key" -in "$CERTS_DIR/$certName.cert" -keypbe NONE -certpbe NONE -passout pass:
+    
+    # Import .pfx into Key-Vault to be safly stored and used by the application
```
Comment (phongcao) [inline deployment/certs/generate-test-certs-keyvault.sh]: safely

File: deployment/certs/generate-test-certs-keyvault.sh [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+generate_root_ca_cert() {
+    local certName=$1
+
+    ###########################
+    # Create a custom CA certificate to be used for 
+    # signing both cluster and client the certificate
+    ###########################
+
+    # Create custom CA-Certificate in Key-Vault
+    #CSR=$(az keyvault certificate create --vault-name $keyVault --name $certName --policy "@$certName-cert-policy.json" | jq -r '.csr')
```
Comment (phongcao) [inline deployment/certs/generate-test-certs-keyvault.sh]: If we can generate the custom ca and sign the certs using az cli, that would be perfect right? Maybe this is something we can investigate in the future.

**Review by phongcao — APPROVED** [review]

File: deployment/certs/generate-test-certs-keyvault.sh (hunk 7 above)
Comment (phongcao) [reply deployment/certs/generate-test-certs-keyvault.sh]: Maybe we can create an open issue to keep track of this issue.

Comment (Aymalla) [reply deployment/certs/generate-test-certs-keyvault.sh]: Issue has been created. #4

File: deployment/certs/generate-test-certs-keyvault.sh (hunk 6 above)
Comment (Aymalla) [reply deployment/certs/generate-test-certs-keyvault.sh]: Sorted

File: temporal-mtls.md (hunk 5 above)
Comment (Aymalla) [reply temporal-mtls.md:1]: Done

**Pull request merged**